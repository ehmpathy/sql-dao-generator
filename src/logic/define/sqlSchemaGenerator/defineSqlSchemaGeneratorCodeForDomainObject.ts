import { camelCase, noCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { isPresent } from 'simple-type-guards';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UserInputError } from '../../UserInputError';
import { isNotADatabaseGeneratedProperty } from '../sqlSchemaRelationship/isNotADatabaseGeneratedProperty';
import { defineSqlSchemaGeneratorCodeForProperty } from './defineSqlSchemaGeneratorCodeForProperty';

export const defineSqlSchemaGeneratorCodeForDomainObject = ({
  domainObject,
  sqlSchemaRelationship,
}: {
  domainObject: DomainObjectMetadata;
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
}) => {
  // define which sql-schema-generator class we'll be using for defining the schema for this domain object
  const schemaGeneratorClass = (() => {
    if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) return 'Entity';
    if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) return 'ValueObject';
    if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) return 'Event';
    throw new UserInputError({
      reason: `sql schema can only be created for a domain object which extends DomainEntity, DomainValueObject, or DomainEvent. Found extends '${domainObject.extends}'`,
      domainObjectName: domainObject.name,
    });
  })();

  // define the code for each property
  const schemaGeneratorProperties = sqlSchemaRelationship.properties
    .filter(isNotADatabaseGeneratedProperty) // filter out all db generated properties, since sql-schema-generator will complain if they're included
    .filter((propertyRelationship) => propertyRelationship.domainObject.name !== 'uuid') // we also can't use the "uuid" property, even if its not db generated, since sql-schema-generator will still complain even if entity is unique on it
    .map((propertyRelationship) =>
      defineSqlSchemaGeneratorCodeForProperty({
        domainObject,
        domainObjectProperty: propertyRelationship.domainObject,
        sqlSchemaProperty: propertyRelationship.sqlSchema,
      }),
    );

  // define the code for the unique properties
  const schemaGeneratorUnique =
    sqlSchemaRelationship.decorations.unique.sqlSchema &&
    domainObject.extends !== DomainObjectVariant.DOMAIN_VALUE_OBJECT
      ? `unique: [${sqlSchemaRelationship.decorations.unique.sqlSchema.map((s) => `'${s}'`).join(', ')}],`
      : null;

  // define the imported props from sql-schema-generator based on this
  const schemaGeneratorImports = [schemaGeneratorClass, 'prop']
    .sort((a, b) => (a.toLowerCase() < b.toLowerCase() ? -1 : 1))
    .join(', ');

  // figure out all of the referenced domain objects and define the imports for those
  const referenceImports = sqlSchemaRelationship.properties
    .map((propertyRelationship) => propertyRelationship.sqlSchema.reference)
    .filter(isPresent)
    .map((reference) => `import { ${camelCase(reference.of.name)} } from './${camelCase(reference.of.name)}';`);
  const distinctSortedReferenceImports = [...new Set(referenceImports.sort())];

  // define the code
  const content = `
import { ${schemaGeneratorImports} } from 'sql-schema-generator';
${distinctSortedReferenceImports.length ? ['', ...distinctSortedReferenceImports, ''].join('\n') : ''}
/**
 * sql-schema for the ${noCase(domainObject.extends)} '${domainObject.name}'
 */
export const ${camelCase(domainObject.name)} = new ${schemaGeneratorClass}({
  name: '${sqlSchemaRelationship.name.sqlSchema}',
  properties: {
    ${schemaGeneratorProperties.join('\n    ')}
  },
${schemaGeneratorUnique ? `  ${schemaGeneratorUnique}\n` : ''}});
  `.trim();

  return content;
};
