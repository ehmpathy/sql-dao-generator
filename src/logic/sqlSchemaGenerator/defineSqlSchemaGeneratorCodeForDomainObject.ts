import { camelCase, noCase, snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';

import { defineSqlSchemaGeneratorPropertyForDomainObjectProperty } from './defineSqlSchemaGeneratorPropertyForDomainObjectProperty';

export const defineSqlSchemaGeneratorCodeForDomainObject = ({
  metadata,
  allMetadatas,
}: {
  metadata: DomainObjectMetadata;
  allMetadatas: DomainObjectMetadata[];
}) => {
  // define which sql-schema-generator class we'll be using for defining the schema for this domain object
  const schemaGeneratorClass = (() => {
    if (metadata.extends === DomainObjectVariant.DOMAIN_ENTITY) return 'Entity';
    if (metadata.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) return 'ValueObject';
    if (metadata.extends === DomainObjectVariant.DOMAIN_EVENT) return 'Event';
    throw new Error(
      `sql schema can only be created for a domain object which extends DomainEntity, DomainValueObject, or DomainEvent. '${metadata.name}' does not meet this criteria, as it extends '${metadata.extends}'`,
    );
  })();

  // make sure that entities have "unique" and "updatable" defined
  if (metadata.extends === DomainObjectVariant.DOMAIN_ENTITY) {
    if (!metadata.decorations.unique?.length)
      throw new Error(
        `domain entities must have at least one 'unique' property defined in order for a schema to be generated. '${metadata.name}' does not satisfy this requirement.

        (note: if its not unique on any natural keys, please specify the 'uuid' as it's only unique property. e.g., \`['uuid']\`) `,
      );
    if (!Array.isArray(metadata.decorations.updatable))
      throw new Error(`domain entities must have their 'updatable' properties defined in order for a schema to be generated. '${metadata.name}' does not satisfy this requirement.

      (note: if it has no updatable properties, please specify this with an empty array. e.g., \`[]\`)`);
  }

  // make sure that value objects do not have unique or updatable defined
  if (metadata.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) {
    if (metadata.decorations.unique)
      throw new Error(
        `domain value objects must _not_ have their 'unique' properties specified. value objects are unique on all of their properties by definition. '${metadata.name}' does not satisfy this requirement.`,
      );
    if (metadata.decorations.updatable)
      throw new Error(
        `domain value objects must _not_ have any 'updatable' properties specified. value objects do not have any updatable properties by definition. '${metadata.name}' does not satisfy this requirement.`,
      );
  }

  // make sure that events have "unique" defined
  if (metadata.extends === DomainObjectVariant.DOMAIN_EVENT) {
    if (!metadata.decorations.unique?.length)
      throw new Error(
        `domain events must have at least one 'unique' property defined in order for a schema to be generated. '${metadata.name}' does not satisfy this requirement.`,
      );
  }

  // define the properties of the domain object
  const schemaGeneratorProperties = Object.entries(metadata.properties)
    .filter((entry) => !['id', 'uuid'].includes(entry[0]))
    .map((entry) => {
      const propertyName = entry[0];
      const propertyDefinition = entry[1];
      return defineSqlSchemaGeneratorPropertyForDomainObjectProperty({
        propertyName,
        propertyDefinition,
        metadata,
        allMetadatas,
      });
    });

  // define the unique properties of the schema entity, if any
  const schemaGeneratorUnique =
    metadata.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT
      ? null
      : `unique: [${metadata.decorations.unique!.map((s) => `'${snakeCase(s)}'`).join(', ')}],`;

  // define the imports based on this
  const schemaGeneratorImports = [schemaGeneratorClass, 'prop']
    .sort((a, b) => (a.toLowerCase() < b.toLowerCase() ? -1 : 1))
    .join(', ');
  const content = `
import { ${schemaGeneratorImports} } from 'sql-schema-generator';

/**
 * sql-schema for the ${noCase(metadata.extends)} '${metadata.name}'
 */
export const ${camelCase(metadata.name)} = new ${schemaGeneratorClass}({
  name: '${snakeCase(metadata.name)}',
  properties: {
    ${schemaGeneratorProperties.join('\n    ')}
  },
${schemaGeneratorUnique ? `  ${schemaGeneratorUnique}\n` : ''}});
  `.trim();

  return content;
};
