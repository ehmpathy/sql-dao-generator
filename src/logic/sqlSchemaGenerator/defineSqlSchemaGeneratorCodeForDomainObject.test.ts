import { DomainObjectMetadata, DomainObjectPropertyType, DomainObjectVariant } from 'domain-objects-metadata';
import { defineSqlSchemaGeneratorCodeForDomainObject } from './defineSqlSchemaGeneratorCodeForDomainObject';

describe('defineSqlSchemaGeneratorCodeForDomainObject', () => {
  describe('names', () => {
    it('should define the generator object name in camel case', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'GreatStuff',
          extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
          properties: {
            latitude: {
              type: DomainObjectPropertyType.NUMBER,
            },
            longitude: {
              type: DomainObjectPropertyType.NUMBER,
            },
          },
          decorations: {
            unique: null,
            updatable: null,
          },
        }),
        allMetadatas: [],
      });
      expect(code).toContain('export const greatStuff = new '); // camel case
    });
    it('should define the table name in snake case', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'GreatStuff',
          extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
          properties: {
            latitude: {
              type: DomainObjectPropertyType.NUMBER,
            },
            longitude: {
              type: DomainObjectPropertyType.NUMBER,
            },
          },
          decorations: {
            unique: null,
            updatable: null,
          },
        }),
        allMetadatas: [],
      });
      expect(code).toContain("name: 'great_stuff'"); // snake case
    });
    it('should define property names in snake case', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            socialSecurityNumberHash: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            name: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            hasCertificate: {
              type: DomainObjectPropertyType.BOOLEAN,
              nullable: true,
            },
          },
          decorations: {
            unique: ['socialSecurityNumberHash'],
            updatable: ['name', 'hasCertificate'],
          },
        }),
        allMetadatas: [],
      });
      expect(code).toContain('social_security_number_hash: '); // snake
      expect(code).toContain('has_certificate: '); // snake
      expect(code).toMatchSnapshot(); // to save an example
    });
    it('should add the _id suffix to directly nested reference property names', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            socialSecurityNumberHash: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            name: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            steamEngineCertificate: {
              type: DomainObjectPropertyType.REFERENCE,
              of: {
                name: 'Certificate',
                extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
              },
            },
          },
          decorations: {
            unique: ['socialSecurityNumberHash'],
            updatable: ['name', 'certificate'],
          },
        }),
        allMetadatas: [],
      });
      // console.log(code);
      expect(code).toContain('steam_engine_certificate_id: '); // snake
      expect(code).toMatchSnapshot(); // to save an example
    });
    it('should add the _ids suffix and strip plurality to directly nested reference array property names', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            socialSecurityNumberHash: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            name: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            engineCertificates: {
              type: DomainObjectPropertyType.ARRAY,
              of: {
                type: DomainObjectPropertyType.REFERENCE,
                of: {
                  name: 'Certificate',
                  extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
                },
              },
            },
          },
          decorations: {
            unique: ['socialSecurityNumberHash'],
            updatable: ['name', 'engineCertificates'],
          },
        }),
        allMetadatas: [],
      });
      // console.log(code);
      expect(code).toContain('engine_certificate_ids: '); // snake
      expect(code).toMatchSnapshot(); // to save an example
    });
    it('should add the _id suffix to inferred uuid reference property names', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            socialSecurityNumberHash: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            name: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            steamEngineCertificateUuid: {
              type: DomainObjectPropertyType.STRING,
            },
          },
          decorations: {
            unique: ['socialSecurityNumberHash'],
            updatable: ['name', 'steamEngineCertificateUuid'],
          },
        }),
        allMetadatas: [
          {
            name: 'Certificate',
            extends: DomainObjectVariant.DOMAIN_ENTITY,
          },
        ] as DomainObjectMetadata[],
      });
      // console.log(code);
      expect(code).toContain('steam_engine_certificate_id: '); // snake
      expect(code).toMatchSnapshot(); // to save an example
    });
    it('should add the _ids suffix and strip plurality to inferred uuid reference array property names', () => {
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        metadata: new DomainObjectMetadata({
          name: 'TrainEngineer',
          extends: DomainObjectVariant.DOMAIN_ENTITY,
          properties: {
            socialSecurityNumberHash: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            name: {
              type: DomainObjectPropertyType.STRING,
              required: true,
            },
            engineCertificateUuids: {
              type: DomainObjectPropertyType.ARRAY,
              of: {
                type: DomainObjectPropertyType.STRING,
              },
            },
          },
          decorations: {
            unique: ['socialSecurityNumberHash'],
            updatable: ['name', 'engineCertificateUuids'],
          },
        }),
        allMetadatas: [
          {
            name: 'Certificate',
            extends: DomainObjectVariant.DOMAIN_ENTITY,
          },
        ] as DomainObjectMetadata[],
      });
      // console.log(code);
      expect(code).toContain('engine_certificate_ids: '); // snake
      expect(code).toMatchSnapshot(); // to save an example
    });
  });
  describe('domain object variants', () => {
    describe('domain value object', () => {
      it('should throw an error if unique properties are attempted to be defined for domain-value-object', () => {
        try {
          defineSqlSchemaGeneratorCodeForDomainObject({
            metadata: new DomainObjectMetadata({
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
              properties: {
                latitude: {
                  type: DomainObjectPropertyType.NUMBER,
                },
                longitude: {
                  type: DomainObjectPropertyType.NUMBER,
                },
              },
              decorations: {
                unique: ['latitude'],
                updatable: null,
              },
            }),
            allMetadatas: [],
          });
          throw new Error('should not reach here');
        } catch (error) {
          expect(error.message).toContain("domain value objects must _not_ have their 'unique' properties specified");
        }
      });
      it('should throw an error if updatable properties are attempted to be defined for domain-value-object', () => {
        try {
          defineSqlSchemaGeneratorCodeForDomainObject({
            metadata: new DomainObjectMetadata({
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
              properties: {
                latitude: {
                  type: DomainObjectPropertyType.NUMBER,
                },
                longitude: {
                  type: DomainObjectPropertyType.NUMBER,
                },
              },
              decorations: {
                unique: null,
                updatable: ['latitude'],
              },
            }),
            allMetadatas: [],
          });
          throw new Error('should not reach here');
        } catch (error) {
          expect(error.message).toContain("domain value objects must _not_ have any 'updatable' properties specified");
        }
      });
      it('should create a correct looking sql-schema-generator object for a domain-value-object', () => {
        const code = defineSqlSchemaGeneratorCodeForDomainObject({
          metadata: new DomainObjectMetadata({
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
            properties: {
              id: {
                type: DomainObjectPropertyType.NUMBER,
                required: false,
              },
              uuid: {
                type: DomainObjectPropertyType.STRING,
                required: false,
              },
              latitude: {
                type: DomainObjectPropertyType.NUMBER,
              },
              longitude: {
                type: DomainObjectPropertyType.NUMBER,
              },
            },
            decorations: {
              unique: null,
              updatable: null,
            },
          }),
          allMetadatas: [],
        });
        expect(code).toContain("import { prop, ValueObject } from 'sql-schema-generator'");
        expect(code).toContain('export const geocode = new ValueObject({');
        expect(code).toContain("name: 'geocode'");
        expect(code).not.toContain('id:');
        expect(code).not.toContain('uuid:');
        expect(code).toContain('latitude: prop.NUMERIC()');
        expect(code).toContain('longitude: prop.NUMERIC()');
        expect(code).not.toContain('unique:');
        expect(code).toMatchSnapshot(); // log example
      });
    });
    describe('domain entity', () => {
      it('should throw an error if domain-entity does not have unique properties defined', () => {
        try {
          defineSqlSchemaGeneratorCodeForDomainObject({
            metadata: new DomainObjectMetadata({
              name: 'Carriage',
              extends: DomainObjectVariant.DOMAIN_ENTITY,
              properties: {
                id: {
                  type: DomainObjectPropertyType.NUMBER,
                  required: false,
                },
                uuid: {
                  type: DomainObjectPropertyType.STRING,
                  required: false,
                },
                cin: {
                  type: DomainObjectPropertyType.STRING,
                  required: true,
                },
                carries: {
                  type: DomainObjectPropertyType.ENUM,
                  of: ['PASSENGER', 'FREIGHT'],
                  required: true,
                },
                capacity: {
                  type: DomainObjectPropertyType.NUMBER,
                  nullable: true,
                },
              },
              decorations: {
                unique: null,
                updatable: [],
              },
            }),
            allMetadatas: [],
          });
          throw new Error('should not reach here');
        } catch (error) {
          expect(error.message).toContain(
            "domain entities must have at least one 'unique' property defined in order for a schema to be generated",
          );
        }
      });
      it('should throw an error if domain-entity does not have updatable properties defined', () => {
        try {
          defineSqlSchemaGeneratorCodeForDomainObject({
            metadata: new DomainObjectMetadata({
              name: 'Carriage',
              extends: DomainObjectVariant.DOMAIN_ENTITY,
              properties: {
                id: {
                  type: DomainObjectPropertyType.NUMBER,
                  required: false,
                },
                uuid: {
                  type: DomainObjectPropertyType.STRING,
                  required: false,
                },
                cin: {
                  type: DomainObjectPropertyType.STRING,
                  required: true,
                },
                carries: {
                  type: DomainObjectPropertyType.ENUM,
                  of: ['PASSENGER', 'FREIGHT'],
                  required: true,
                },
                capacity: {
                  type: DomainObjectPropertyType.NUMBER,
                  nullable: true,
                },
              },
              decorations: {
                unique: ['cin'],
                updatable: null,
              },
            }),
            allMetadatas: [],
          });
          throw new Error('should not reach here');
        } catch (error) {
          expect(error.message).toContain(
            "domain entities must have their 'updatable' properties defined in order for a schema to be generated.",
          );
        }
      });
      it('should create a correct looking sql-schema-generator object for a domain-entity', () => {
        const code = defineSqlSchemaGeneratorCodeForDomainObject({
          metadata: new DomainObjectMetadata({
            name: 'Carriage',
            extends: DomainObjectVariant.DOMAIN_ENTITY,
            properties: {
              id: {
                type: DomainObjectPropertyType.NUMBER,
                required: false,
              },
              uuid: {
                type: DomainObjectPropertyType.STRING,
                required: false,
              },
              cin: {
                type: DomainObjectPropertyType.STRING,
                required: true,
              },
              carries: {
                type: DomainObjectPropertyType.ENUM,
                of: ['PASSENGER', 'FREIGHT'],
                required: true,
              },
              capacity: {
                type: DomainObjectPropertyType.NUMBER,
                nullable: true,
              },
            },
            decorations: {
              unique: ['cin'],
              updatable: [],
            },
          }),
          allMetadatas: [],
        });
        expect(code).toContain("import { Entity, prop } from 'sql-schema-generator'");
        expect(code).toContain('export const carriage = new Entity({');
        expect(code).toContain("name: 'carriage'");
        expect(code).not.toContain('id:'); // should be filtered out
        expect(code).not.toContain('uuid:'); // should be filtered out
        expect(code).toContain('cin: prop.VARCHAR(),');
        expect(code).toContain("carries: prop.ENUM(['PASSENGER', 'FREIGHT']),");
        expect(code).toContain('capacity: { ...prop.NUMERIC(), nullable: true },');
        expect(code).toContain("unique: ['cin'],");
        expect(code).toMatchSnapshot();
      });
    });
    describe('domain event', () => {
      it('should throw an error if unique properties are not defined for a domain-event', () => {
        try {
          defineSqlSchemaGeneratorCodeForDomainObject({
            metadata: new DomainObjectMetadata({
              name: 'TrainLocatedEvent',
              extends: DomainObjectVariant.DOMAIN_EVENT,
              properties: {
                id: {
                  type: DomainObjectPropertyType.NUMBER,
                  required: false,
                },
                trainUuid: {
                  type: DomainObjectPropertyType.STRING,
                  required: true,
                },
                occurredAt: {
                  type: DomainObjectPropertyType.DATE,
                  required: true,
                },
                geocode: {
                  type: DomainObjectPropertyType.REFERENCE,
                  of: {
                    name: 'Geocode',
                    extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
                  },
                  required: true,
                },
              },
              decorations: {
                unique: null,
                updatable: [],
              },
            }),
            allMetadatas: [],
          });
          throw new Error('should not reach here');
        } catch (error) {
          expect(error.message).toContain(
            "domain events must have at least one 'unique' property defined in order for a schema to be generated.",
          );
        }
      });
      it('should create a correct looking sql-schema-generator object for a domain-event', () => {
        const code = defineSqlSchemaGeneratorCodeForDomainObject({
          metadata: new DomainObjectMetadata({
            name: 'TrainLocatedEvent',
            extends: DomainObjectVariant.DOMAIN_EVENT,
            properties: {
              id: {
                type: DomainObjectPropertyType.NUMBER,
                required: false,
              },
              trainUuid: {
                type: DomainObjectPropertyType.STRING,
                required: true,
              },
              occurredAt: {
                type: DomainObjectPropertyType.DATE,
                required: true,
              },
              geocode: {
                type: DomainObjectPropertyType.REFERENCE,
                of: {
                  name: 'Geocode',
                  extends: DomainObjectVariant.DOMAIN_VALUE_OBJECT,
                },
                required: true,
              },
            },
            decorations: {
              unique: ['trainUuid', 'occurredAt'],
              updatable: [],
            },
          }),
          allMetadatas: [],
        });
        expect(code).toContain("import { Event, prop } from 'sql-schema-generator'");
        expect(code).toContain('export const trainLocatedEvent = new Event({');
        expect(code).toContain("name: 'train_located_event'");
        expect(code).not.toContain(' id:'); // should be filtered out
        expect(code).not.toContain(' uuid:'); // should be filtered out
        expect(code).toContain('train_uuid: prop.VARCHAR(),');
        expect(code).toContain('occurred_at: prop.TIMESTAMPTZ(),');
        expect(code).toContain('geocode_id: prop.REFERENCES(geocode),');
        expect(code).toContain("unique: ['train_uuid', 'occurred_at'],");
        expect(code).toMatchSnapshot();
      });
    });
  });
  describe('unique', () => {
    // should be correct for null
    // should be correct for explicit one
    // should be correct for explicit more than one
    // should be correct for only on uuid
  });
});
