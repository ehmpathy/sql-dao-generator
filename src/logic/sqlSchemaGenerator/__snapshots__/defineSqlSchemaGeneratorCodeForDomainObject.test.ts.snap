// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants domain entity should create a correct looking sql-schema-generator object for a domain-entity 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'Carriage'
 */
export const carriage = new Entity({
  name: 'carriage',
  properties: {
    cin: prop.VARCHAR(),
    carries: prop.ENUM(['PASSENGER', 'FREIGHT']),
    capacity: { ...prop.NUMERIC(), nullable: true },
  },
  unique: ['cin'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants domain event should create a correct looking sql-schema-generator object for a domain-event 1`] = `
"import { Event, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain event 'TrainLocatedEvent'
 */
export const trainLocatedEvent = new Event({
  name: 'train_located_event',
  properties: {
    train_uuid: prop.VARCHAR(),
    occurred_at: prop.TIMESTAMPTZ(),
    geocode_id: prop.REFERENCES(geocode),
  },
  unique: ['train_uuid', 'occurred_at'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants domain value object should create a correct looking sql-schema-generator object for a domain-value-object 1`] = `
"import { prop, ValueObject } from 'sql-schema-generator';

/**
 * sql-schema for the domain value object 'Geocode'
 */
export const geocode = new ValueObject({
  name: 'geocode',
  properties: {
    latitude: prop.NUMERIC(),
    longitude: prop.NUMERIC(),
  },
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject names should add the _id suffix to directly nested reference property names 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'TrainEngineer'
 */
export const trainEngineer = new Entity({
  name: 'train_engineer',
  properties: {
    social_security_number_hash: prop.VARCHAR(),
    name: { ...prop.VARCHAR(), updatable: true },
    steam_engine_certificate_id: prop.REFERENCES(certificate),
  },
  unique: ['social_security_number_hash'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject names should add the _id suffix to inferred uuid reference property names 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'TrainEngineer'
 */
export const trainEngineer = new Entity({
  name: 'train_engineer',
  properties: {
    social_security_number_hash: prop.VARCHAR(),
    name: { ...prop.VARCHAR(), updatable: true },
    steam_engine_certificate_id: { ...prop.REFERENCES(certificate), updatable: true },
  },
  unique: ['social_security_number_hash'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject names should add the _ids suffix and strip plurality to directly nested reference array property names 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'TrainEngineer'
 */
export const trainEngineer = new Entity({
  name: 'train_engineer',
  properties: {
    social_security_number_hash: prop.VARCHAR(),
    name: { ...prop.VARCHAR(), updatable: true },
    engine_certificate_ids: { ...prop.ARRAY_OF(prop.REFERENCES(certificate)), updatable: true },
  },
  unique: ['social_security_number_hash'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject names should add the _ids suffix and strip plurality to inferred uuid reference array property names 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'TrainEngineer'
 */
export const trainEngineer = new Entity({
  name: 'train_engineer',
  properties: {
    social_security_number_hash: prop.VARCHAR(),
    name: { ...prop.VARCHAR(), updatable: true },
    engine_certificate_ids: { ...prop.ARRAY_OF(prop.REFERENCES(certificate)), updatable: true },
  },
  unique: ['social_security_number_hash'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject names should define property names in snake case 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'TrainEngineer'
 */
export const trainEngineer = new Entity({
  name: 'train_engineer',
  properties: {
    social_security_number_hash: prop.VARCHAR(),
    name: { ...prop.VARCHAR(), updatable: true },
    has_certificate: { ...prop.BOOLEAN(), updatable: true, nullable: true },
  },
  unique: ['social_security_number_hash'],
});"
`;
